### PR TITLE
added Button to Open Circuit Link in Browser

### LINKED ISSUE #261: Add Button To Open Circuit Link In Browser
**Is your feature request related to a problem? Please describe.**
Whenever any user is visiting an particular circuit , we have option to share but no option to open the link in browser. The browser can show the circuit preview easily.

**How to get this page ?**
1. Open Sidebar
2. Click on "Explore"
3. Click on "Featured Circuits"
4. Tap on any circuit

**Describe the solution you'd like**
We can put a button on the navbar at the side of the share button . On click this button that will open the link in browser.
For just a little explanation, see this image

**Are you working on this? (Yes/No)**
No

### PR DESCRIPTION
**Pull request by sainideepanshu199**
Fixes #261 

Describe the changes you have made in this PR -
added button that opens the circuit link in device default browser

Screenshots of the changes (If any) -

Screenshot_20230227-234119 1
Screenshot_20230227-234116 1

Note: Please check Allow edits from maintainers. if you would like us to assist in the PR.

### CODE REVIEW

File: lib/ui/views/projects/project_details_view.dart [hunk 1]
```diff
@@ -64,6 +66,20 @@ class _ProjectDetailsViewState extends State<ProjectDetailsView> {
...
   }
 
+  Widget _buildOpenActionButton() {
+    return Padding(
+      padding: const EdgeInsets.symmetric(horizontal: 8),
+      child: IconButton(
+        onPressed: () async {
+          final url = Uri.parse(
+              'https://circuitverse.org/users/${widget.project.relationships.author.data.id}/projects/${widget.project.id}');
+          await launchUrl(url, mode: LaunchMode.externalApplication);
```
Comment (Tanmoy741127) [inline lib/ui/views/projects/project_details_view.dart:76]: There is already an predefined function for launchUrl in lib/utils/url_launcher.dart, kindly use that function to open the link.

Comment (sainideepanshu199) [reply lib/ui/views/projects/project_details_view.dart]: i have changed the code accordingly , please review the code.

**Review by Tanmoy741127 — APPROVED** [review]

**Comment by Tanmoy741127** [comment]
@vedant-jain03 it's ready to be merged

**Review by vedant-jain03 — APPROVED** [review]
cc @aman-singh7 @manjotsidhu

**Comment by Tanmoy741127** [comment]
> @Tanmoy741127 @sainideepanshu199 Here as well launchURL function could be used in alot of places, we should not be causing effects in other parts of the apps.

ok !

File: lib/utils/url_launcher.dart [hunk 2]
```diff
@@ -2,7 +2,7 @@ import 'package:url_launcher/url_launcher_string.dart';
 
 void launchURL(String url) async {
   if (await canLaunchUrlString(url)) {
-    await launchUrlString(url);
+    await launchUrlString(url, mode: LaunchMode.externalApplication);
```
Comment (Tanmoy741127) [inline lib/utils/url_launcher.dart:5]: By default , keep it launchUrlString(url) without  mode: LaunchMode.externalApplication

Pass an varibale in function `openInExternalApplication` with default value `False`

Do the necessary changes, in the ` mode: LaunchMode.externalApplication` also.

**Review by Tanmoy741127 — CHANGES_REQUESTED** [review]

File: lib/utils/url_launcher.dart (hunk 2 above)
Comment (sainideepanshu199) [reply lib/utils/url_launcher.dart:5]: i have made the changes in pr#270 , so now only specific links will open with external browser, rest all will open in app web view

**Comment by sainideepanshu199** [comment]
i am getting errors in some files , can anyone tell why is this error 
image

**Comment by Tanmoy741127** [comment]
> i am getting errors in some files , can anyone tell why is this error image

The issue due to some unresolved third party package . Run `flutter pub get` and check whether the issue got fixed.

**Comment by sainideepanshu199** [comment]
> > i am getting errors in some files , can anyone tell why is this error image
> 
> The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.

no even after running flutter pub get , i am getting same issue

**Comment by Tanmoy741127** [comment]
> > > i am getting errors in some files , can anyone tell why is this error image
> > 
> > 
> > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> 
> no even after running flutter pub get , i am getting same issue

Can you hover on the issue and check whether it's a import issue or deprecation warning

**Comment by sainideepanshu199** [comment]
> > > > i am getting errors in some files , can anyone tell why is this error image
> > > 
> > > 
> > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > 
> > 
> > no even after running flutter pub get , i am getting same issue
> 
> Can you hover on the issue and check whether it's a import issue or deprecation warning

mostly errors are due to argument cannot be assigned to parameter

**Comment by Tanmoy741127** [comment]
> > > > > i am getting errors in some files , can anyone tell why is this error image
> > > > 
> > > > 
> > > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > > 
> > > 
> > > no even after running flutter pub get , i am getting same issue
> > 
> > 
> > Can you hover on the issue and check whether it's a import issue or deprecation warning
> 
> mostly errors are due to argument cannot be assigned to parameter

Ok I will check

**Comment by sainideepanshu199** [comment]
> > > > > > i am getting errors in some files , can anyone tell why is this error image
> > > > > 
> > > > > 
> > > > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > > > 
> > > > 
> > > > no even after running flutter pub get , i am getting same issue
> > > 
> > > 
> > > Can you hover on the issue and check whether it's a import issue or deprecation warning
> > 
> > 
> > mostly errors are due to argument cannot be assigned to parameter
> 
> Ok I will check

is their any update on these errors as i am still getting these errors and it's giving problem in the build and development of app. please tell if i am the only one facing the problem.

**Comment by Tanmoy741127** [comment]
> > > > > > > i am getting errors in some files , can anyone tell why is this error image
> > > > > > 
> > > > > > 
> > > > > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > > > > 
> > > > > 
> > > > > no even after running flutter pub get , i am getting same issue
> > > > 
> > > > 
> > > > Can you hover on the issue and check whether it's a import issue or deprecation warning
> > > 
> > > 
> > > mostly errors are due to argument cannot be assigned to parameter
> > 
> > 
> > Ok I will check
> 
> is their any update on these errors as i am still getting these errors and it's giving problem in the build and development of app. please tell if i am the only one facing the problem.

I haven't got time to check, will trying to check today

**Comment by Tanmoy741127** [comment]
> > > > > > > i am getting errors in some files , can anyone tell why is this error image
> > > > > > 
> > > > > > 
> > > > > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > > > > 
> > > > > 
> > > > > no even after running flutter pub get , i am getting same issue
> > > > 
> > > > 
> > > > Can you hover on the issue and check whether it's a import issue or deprecation warning
> > > 
> > > 
> > > mostly errors are due to argument cannot be assigned to parameter
> > 
> > 
> > Ok I will check
> 
> is their any update on these errors as i am still getting these errors and it's giving problem in the build and development of app. please tell if i am the only one facing the problem.

I have checked in my local system, I have not that issue you said.

**Comment by sainideepanshu199** [comment]
> > > > > > > > i am getting errors in some files , can anyone tell why is this error image
> > > > > > > 
> > > > > > > 
> > > > > > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > > > > > 
> > > > > > 
> > > > > > no even after running flutter pub get , i am getting same issue
> > > > > 
> > > > > 
> > > > > Can you hover on the issue and check whether it's a import issue or deprecation warning
> > > > 
> > > > 
> > > > mostly errors are due to argument cannot be assigned to parameter
> > > 
> > > 
> > > Ok I will check
> > 
> > 
> > is their any update on these errors as i am still getting these errors and it's giving problem in the build and development of app. please tell if i am the only one facing the problem.
> 
> I have checked in my local system, I have not that issue you said.

what shall i do then

**Comment by Tanmoy741127** [comment]
> > > > > > > > > i am getting errors in some files , can anyone tell why is this error image
> > > > > > > > 
> > > > > > > > 
> > > > > > > > The issue due to some unresolved third party package import maybe. Run `flutter pub get` and check whether the issue got fixed.
> > > > > > > 
> > > > > > > 
> > > > > > > no even after running flutter pub get , i am getting same issue
> > > > > > 
> > > > > > 
> > > > > > Can you hover on the issue and check whether it's a import issue or deprecation warning
> > > > > 
> > > > > 
> > > > > mostly errors are due to argument cannot be assigned to parameter
> > > > 
> > > > 
> > > > Ok I will check
> > > 
> > > 
> > > is their any update on these errors as i am still getting these errors and it's giving problem in the build and development of app. please tell if i am the only one facing the problem.
> > 
> > 
> > I have checked in my local system, I have not that issue you said.
> 
> what shall i do then

Try to run/build the app . Sometime the errors show due to vscode issues.

**Comment by sainideepanshu199** [comment]
can anyone tell the status of this pr, does it has errors or is it not reviewed till now.

**Pull request closed without merging**